cajita/src/Cajita_SparseDimPartitioner.hpp

```diff
@@ -0,0 +1,687 @@
+            tiles_per_dim[i] <<= cell_bits_per_tile_dim * 3;
+        }
+        return tiles_per_dim;
+    }
+
+    /*!
+      \brief Initialize the tile partition; partition in each dimension
+      has the form [0, p_1, ..., p_n, total_tile_num], so the partition
+      would be [0, p_1), [p_1, p_2) ... [p_n, total_tile_num]
+      \param rec_partition_i partition array in dimension i
+      \param rec_partition_j partition array in dimension j
+      \param rec_partition_k partition array in dimension k
+    */
+    void initializeRecPartition( std::vector<int>& rec_partition_i,
+                                 std::vector<int>& rec_partition_j,
+                                 std::vector<int>& rec_partition_k )
+    {
+
+        int max_size = 0;
+        for ( int d = 0; d < 3; ++d )
+            max_size =
+                max_size < _ranks_per_dim[d] ? _ranks_per_dim[d] : max_size;
+
+        typedef typename execution_space::array_layout layout;
+        Kokkos::View<int* [3], layout, Kokkos::HostSpace> rectangle_partition(
+            "rectangle_partition_host", max_size + 1 );
+
+        for ( int id = 0; id < _ranks_per_dim[0] + 1; ++id )
+            rectangle_partition( id, 0 ) = rec_partition_i[id];
+
+        for ( int id = 0; id < _ranks_per_dim[1] + 1; ++id )
+            rectangle_partition( id, 1 ) = rec_partition_j[id];
+
+        for ( int id = 0; id < _ranks_per_dim[2] + 1; ++id )
+            rectangle_partition( id, 2 ) = rec_partition_k[id];
+
+        _rectangle_partition_dev = Kokkos::create_mirror_view_and_copy(
+            memory_space(), rectangle_partition );
+    }
+
+    /*!
+      \brief Get the current partition.
+      Copy partition from the device view to host std::array<vector>
+    */
+    std::array<std::vector<int>, 3> getCurrentPartition()
+    {
+        std::array<std::vector<int>, 3> rec_part;
+        auto rec_mirror = Kokkos::create_mirror_view_and_copy(
+            Kokkos::HostSpace(), _rectangle_partition_dev );
+        for ( int d = 0; d < 3; ++d )
+        {
+            rec_part[d].resize( _ranks_per_dim[d] );
+            for ( int id = 0; id < _ranks_per_dim[d] + 1; ++id )
+            {
+                rec_part[d][id] = rec_mirror( id, d );
+            }
+        }
+        return rec_part;
+    }
+
+    /*!
+         \brief set all elements in _workload_per_tile and _workload_prefix_sum
+         matrix to 0
+       */
+    void resetWorkload()
+    {
+        Kokkos::deep_copy( _workload_per_tile, 0 );
+        Kokkos::deep_copy( _workload_prefix_sum, 0 );
+    }
+
+    /*!
+      \brief compute the workload in the current MPI rank from particle
+      positions (each particle count for 1 workload value)
+      \param view particle positions view
+      \param particle_num total particle number
+      \param dx cell dx size
+    */
+    template <class ParticlePosViewType, typename CellUnit>
+    void computeLocalWorkLoad( ParticlePosViewType& view, int particle_num,
```

Comment:
I don't think we're testing the particle workload codepath - is that relatively easy to add? Or should we leave that until later (if so maybe we remove this for now)?